### PR TITLE
add support for Jetbrains Toolbox

### PR DESCRIPTION
**Pull request by FalkoHilbert**
This PR adds the ability to automatically detect the most recent version installed by the Jetbrain Toolbox.
In addition, the code automatically determines the path to the installtions folder of the installations through the toolbox.
With this code, the user only has to set the 'toolBoxActive' flag to true and the path as well as the version name and window name are automatically determined.

### CODE REVIEW

**Comment by FalkoHilbert** [comment]
> Can you please explain what's actually is done by the added code in PR description?
> 
> I'm not using JetBrains Toolbox myself, but it's it just a way to automatically download JetBrains products. At the end only folder where PhpStorm is located is changed and you already can specify that folder at script beginning.

done ✔️

**Comment by aik099** [comment]
Nice. Could you refactor a code to use functions or classes maybe? Right now it looks hard to maintain.

**Comment by FalkoHilbert** [comment]
Of course, I would outsource the appropriate code in a function, but not more. Since I'm not so fit with Microsoft's JScript.

**Comment by FalkoHilbert** [comment]
i have moved the code and added some comments. I think that more refactoring is a bit oversized.

File: PhpStorm Protocol (Win)/run_editor.js [hunk 1]
```diff
@@ -1,57 +1,117 @@
...
+
+
+// don't change anything below this line, unless you know what you're doing
+var	url = WScript.Arguments(0),
+	match = /^phpstorm:\/\/open\/?\?(url=file:\/\/|file=)(.+)&line=(\d+)$/.exec(url),
+	project = '',
+	editor = '"C:\\' + (settings.x64 ? 'Program Files' : 'Program Files (x86)') + '\\JetBrains\\' + settings.folder_name + (settings.x64 ? '\\bin\\phpstorm64.exe' : '\\bin\\phpstorm.exe')+'"';
+
+if( settings.toolBoxActive )
+{
```
Comment (aik099) [inline PhpStorm Protocol (Win)/run_editor.js]: Can you please format this IF statement as another one?

Comment (FalkoHilbert) [reply PhpStorm Protocol (Win)/run_editor.js]: done ✔️

**Comment by aik099** [comment]
Merging, thank you @FalkoHilbert .

**Comment by FalkoHilbert** [comment]
@aik099 could you update the Readme.md? I have just seen that there is nothing in this yet to the feature in it. Somehow I forgot that.

**Comment by aik099** [comment]
@FalkoHilbert , can send new PR that? I have no idea what text you've planned to add there.

**Comment by equistx** [comment]
It is great that you solved this guys, I don't know this lang, or even how to debug it, so I had really hard time to somehow hardcode paths and window title in there in the past, that took me a while.

Unfurtunately, it is not working (anymore?). It is not able to find newest version so it opens old one instead. In my case:
- old version 2019.2, folder name 192.6262.66
- new version 2019.3, folder name 193.5233.101

I found out, that when I change `5` after first dot to something bigger than `6` ie. `7`, then it starts working as expected. 

version

**Comment by equistx** [comment]
Ok, so issue is in these conditions. It finds newer major `193` than `192`, but next condition is not met, because minor `5233` is not newer than `6262`, `maxVersionFolder` get updated only in the innermost condition, therefore it points to bad version. See image - I "fixed" it by resetting parts when something change. But I would suggest to use some tested semver cmp function, I'm not 100 % sure if it's right like this.

Now, please, tell me, how do you debug this script? Is there any debugger for it? In the end I've used powershell `cscript.exe` command and `WScript.echo()` inside code to debug this thing. But it is debugging nightmare, is there any better way? Excuse me please, I've never written any non browser javascript.

image

**Comment by aik099** [comment]
> But I would suggest to use some tested semver cmp function ...

@equistx , sure. Here is some random function, that looks promising: https://helloacm.com/the-javascript-function-to-compare-version-number-strings/. The idea: you store matched folder name in `x.y.z` format in `maxVersion` variable instead of having separate `maxMajor`, `maxMinor`, `maxPatch` variables and just compare 2 folder names (from `maxVersion` and current folder name) using `compareVersion` function above shown page. Also handle, when `maxVersion` variable is empty and should be set to 1st matched folder.

> Now, please, tell me, how do you debug this script? Is there any debugger for it? In the end I've used powershell cscript.exe command and WScript.echo() inside code to debug this thing.

I've found https://www.robvanderwoude.com/vbstech_debugging.php article about VBScript debugging (this is the language used in that JS file). According to it there are several real-time debugging tools available (some of them are likely completely free):

* Microsoft Script Debugger
* Microsoft Script Editor (comes with Microsoft Office)
* Visual Studio's debugger

To start debugging run this command `CSCRIPT //X yourscript.vbs`.

**Comment by FalkoHilbert** [comment]
@equistx thanks for the Bug report and fix.
I'll make a PR as soon as possible.

**Pull request merged**